replace COMPUTE, RANDOM and applyVariables by operation strings and inline variables

ArnoldSmith86 opened the pull request:
This is just an idea I had a long time ago that came back to me when looking at @larkob's #305. After some discussion in this PR it evolved into this:

**This PR completely eliminates `COMPUTE`, `RANDOM` and `applyVariables`.** They are replaced by expression strings and variables that can be used anywhere.

For an explanation of how to use this, please see [Dynamic Expressions in the wiki](https://github.com/ArnoldSmith86/virtualtabletop/wiki/Dynamic-Expressions-(WIP)).

Here are a few examples of how this changes current routines:

<table>
<tr>
<td> Before </td> <td> After </td>
</tr>
<tr>
<td>

```json
{
  "func": "RANDOM"
}
```
</td>
<td>

```json
"var RANDOM = randInt 1 10"
```
</td>
</tr>
<tr>
<td>

```json
{
  "func": "RANDOM",
  "min": 4,
  "max": 8,
  "variable": "index"
}
```
</td>
<td>

```json
"var index = randInt 4 8"
```
</td>
</tr>
<tr>
<td>

```json
{
  "func": "COMPUTE",
  "operation": "+",
  "operand2": 1,
  "applyVariables": {
    "parameter": "operand1",
    "variable": "index"
  },
  "variable": "index"
}
```
</td>
<td>

```json
"var index = ${index} + 1"
```
</td>
</tr>
<tr>
<td>

```json
{
  "func": "SET",
  "property": "text",
  "applyVariables": {
    "parameter": "value",
    "variable": "name"
  }
}
```
</td>
<td>

```json
{
  "func": "SET",
  "property": "text",
  "value": "${name}"
}
```
</td>
</tr>
</table>

## To do

Initially the plan was for this to exist _in addition_ to `COMPUTE` but I would much prefer to replace it. That means though that the fileupdater needs to migrate everything `COMPUTE` and `applyVariables` can do right now. For the most part that works, but some tricky stuff is still left:

- [x] Strings could contain anything in `COMPUTE` and now they are very limited. The solution will likely be Unicode escapes. They could also be set as widget properties and then used as `${PROPERTY xyz}` but that feels clunky.
- [x] Variable, widget and property names are very restrictive now. Existing ones that don't fit the new format, need to be changed or replaced by an additional `GET` (for widget/property).
- [x] Change some functions (like `max`) to have the parameters after the operation.
- [x] If a `COMPUTE` uses a template in `applyVariables` that is no longer possible with the new syntax. The fileupdater should replace it by a series of `concat` or something like that. In practice no one uses this, I think.
- [x] If a `COMPUTE` has an object or array as one of its operands, the fileupdater needs to work around that.
- [x] If someone uses `${a}` in a string right now, it should be escaped.
- [x] `skip` breaks if the variable was `undefined`.
- [x] The fileupdater doesn't respect it if `variable` was set through `applyVariables`.
- [x] If the `operation` or `variable` are set in `applyVariables` using `property` or `template`, it currently fails.

Other things left to do:

- [x] The top icons in the JSON editor don't align with their hotkey labels anymore because the command pane is narrower now.
- [x] ~~`"var selected = ${pool} getIndex ${index}"` should be transformed to `"var selected = ${pool.$index}"`. I changed my mind because they do behave slightly differently if variables are not set etc. I changed my mind again. I will have a closer look.~~ I changed my mind again. See comment below for why.
- [x] The reset button in Undercover hangs the browser. (It uses an invalid string; probably solved by the string thing above.)
- [x] Allow no operands (`random`) and omitting the operand before the operation (`max 5 8`).
- [x] Add array and object literals (probably only empty).
- [x] Add a message for debug mode if the string didn't match (so it didn't do anything).
- [x] Check if we want to change return values for invalid operations and similar things.
- [x] Remove `RANDOM` as well. It would be part of `COMPUTE` if it didn't predate it.
- [x] JSON editor doesn't detect the correct context for the new expression syntax.
- [x] TestCafé tests need to be updated.
- [x] Setting index `0` seems to be broken.
- [x] Setting `operation` with `applyVariables` is no longer supported.

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3307/pr-test (or any other room on that server)

larkob commented:
This is great! However, round brackets are so commonly used for functions that it's difficult to see them as indicating a widget. I suggest we use angular brackets and add 'w' before the brackets to make this easier to read.

And variables used for either widget IDs or properties could use a dollar sign, so it's relatively easy to differentiate between the use of an ID and and a variable.

```
w[a1b2.x] = 2 //property x of widget with id `a1b2` is set to 2
w[a1b2.$b] = 2 //property defined by variable b of widget with id `a1b2` is set to 2
w[$id.x] = 2 //property defined by variable b of widget with id defined by variable id is set to 2
w[$id.$b] = 2 //property defined by variable b of widget with id defined by variable id is set to 2
a = w[a1b2.x] // set variable a to value of property x of widget `a1b2`
etc.
```

Round brackets could instead be used for compute operations in order to improve readability, with the following syntax:
```
operand1.operation(operand2,operand3)
a = random()
b = 'hello'.charAt(1)
c = $b.length()
d = 'hello'.substr(0,$pos)
e = w[a1b2.x] + 10
```

Characters such as comma, round brackets and dots would be optional, we simply need some delimiters. You could also simply use the space character as a delimiter and write this as, but it just wouldn't as readable:
```
operand1 operation operand2 operand3
a = random
b = 'hello' charAt 1
c = $b length
d = 'hello' substr 0 $pos
e = w[a1b2 x] + 10
```

ArnoldSmith86 commented:
`w[a1b2.x]` doesn't make sense in my mind. It should be `w[a1b2].x` but yeah, the `w[]` might be better.

Your optional delimiter characters will be problematic though. Well probably only the dot because it's used for indexing. You could write `a = charAt.charAt.charAt.charAt` and it would be ambigious. We could ditch the dot for indexing of course and use `var['index']`/`var[var]` or `$var[index]`/`$var[$var]` but I think I prefer the way it is now. That would also mean that setting a property would be something like `$w[$idVar][$indexVar]`.

I was also thinking to add `c` in place of `w` to work with collections. It would enforce that only one collection is used in the whole formula and then you could use `c[DEFAULT].width = c[DEFAULT].width + 100` to add 100 to each widget's width. But that might become too complicated pretty fast and is probably better in a `FOREACH`.

larkob commented:
Yeah, c for collection is interesting. Same with h for holders. We should probably look into this again once we've got a "forEach".

Your example of `a = charAt.charAt.charAt.charAt` shouldn't be possible. Each line should only allow a single operation, which will make reading and parsing much easier. It comes down to this very simple structure:
```
variable = op1 operation op2 op3
```
and
```
widget.property = op1 operation op2 op3
```

Any op could be expressed as a number, a string (using single quotes), a variable (using $prefix) or a widget (using w[] format)

ArnoldSmith86 commented:
Oh, so you want to get rid of the dot for indexing completely? So this would be your suggestion for my example (plus optional delimiter characters)?

```
  "clickRoutine": [
    "$b = 'p'",
    "$a = 1 + 2",
    "w[blvd][p] = $a",
    "w[blvd][p] = w[blvd][p] + w[blvd][$b]",
    "$b = 'asd' split ''",
    "$b[1] = $b[2] concat $b[0]",
    "$c = 'hello' substr 1 2",
    "$var = 'test'",
    "w[blvd][$var] = true",
    "$var2 = 1",
    "$b[$var2] = 99.123"
  ]
```

The `$var[$var]` and `$var[string]` feel weird to me. If it is not `$var.string`, it should be `$var['string']` but I would much prefer the dot. And I don't feel like all those dollar signs are helping readability.

rogerl50 commented:
Another approach is to think of $ as a dereference operator. So for example (returning to using explicit dots, which I prefer as well)
```
"clickRoutine": [
    "b = 'p'",
    "a = 1 + 2",
    "w[blvd].p = $a",
    "w[blvd].p = w[blvd].p + w[blvd].$b",
    "b = 'asd' split ''",
    "b.1 = $b.2 concat $b.0",
    "c = 'hello' substr 1 2",
    "var = 'test'",
    "w[blvd].$var = true",
    "var2 = 1",
    "$b.$var2 = 99.123"
  ]
```
This has the advantage of removing clutter, and allows you to use variables in a more natural way.

A couple of notes:
1. Whatever syntax we end up with, it's important to continue to allow variables to have the same name as properties.
2. Button debug is right now for me at least a must-have when I'm debugging complicated click routines. So whatever we end up with, there has to be a way to support line by line debug output (I notice that the example room already does.)

larkob commented:
As rogerl50 said, I only wanted the dollar sign to be used as a dereference operator. And yes, we definitely need to distinguish widgets from variables. So, if we use `w[id]` then it's unclear whether it refers a widget or a variable named w with an index. And what if the property itself is an array?

A new idea is to introduce "GET" and "SET", which indicates that the following is a widget with property and everyone is already familiar with. 

```
a = random
b = 'hello' charAt 1
c = $b length
d = 'hello' substr 0 $pos
e = GET a1b2.x
f = $arr[2] + 2
g = $arr[$b] + 10
h = GET $widgetId.$propertyName
SET a1b2.x = $a
SET $widgetId.$propertyName = $a + 10
```

With this approach, you couldn't perform to an immediate transformation on a property, but I don't think that's a big deal in contrast to how much this would improve readability.
```
e = a1b2.x + 10 would be written as

e = GET a1b2.x
e = $e + 10
```

ArnoldSmith86 commented:
I think my proposal is to completely remove `COMPUTE` and `applyVariables`, migrating them to this:

```Javascript
  "clickRoutine": [
    "${b} = 'p'",
    "${a} = 1 + 2 // this is a comment", // allow comments in the routine
    "${d} = ${PROPERTY $b OF $idVar}", // the property name and widget ID come from variables
    "${h} = ${PROPERTY p OF blvd} + ${a}", // property name and widget ID are literals (so "id":"blvd")
    "${b} = 'asd' split ''",
    "${b.1} = ${b.2} concat ${b.0}", // variable b, index 0-2
    "${c} = 'hello' substr 1 ${PROPERTY offset}", // uses the property of the widget containing the routine
    "${q} = true",
    "${b.$var2} = 99.123", // the index taken from variable var2
    {
      "func": "MOVE",
      "from": "${var}", // if the string is ONLY a variable, it would not convert the variable contents to a string
      "to": "a template with ${var.index}",
      "count": "${PROPERTY $b}",
      "face": 1
    }
  ]
```

I really prefer the syntax without `${}` of my initial proposal but if we get rid of `applyVariables` and allow the same syntax in operation parameters, I think it really helps that variables look the same everywhere.

I'm pretty sure I actually like that operations that affect the game stand out with their different syntax. That's why I also removed the ability to write to widget properties (reading is fine) so that it wouldn't be a hidden `SET`.

The only functionality this removes (I think) is using `template` in `applyVariables` in `COMPUTE` which - at the time of writing this - is not in use on the production server. I wouldn't want to add template support in the operands because it would make parsing a lot more difficult I think.

robartsd commented:
>     "${d} = ${PROPERTY $b OF $idVar}", // the property name and widget ID come from variables

Should it be "${d} = ${PROPERTY ${b} OF ${idVar}}' for more consistency?

This does eliminate our hacky way of commenting, perhaps we should also add in true comment strings.

ArnoldSmith86 commented:
In theory I agree. I'm not sure if it makes parsing it a nightmare though. And it kinda implies that you can use it recursively..

What do you mean by "true comment strings"? JSON doesn't have _true_ comments.

robartsd commented:
> In theory I agree. I'm not sure if it makes parsing it a nightmare though. And it kinda implies that you can use it recursively..

I understand what you're saying about recursion (and not wanting to support that due to complexity).

> What do you mean by "true comment strings"? JSON doesn't have _true_ comments.

I mean having a syntax to add a string as an element of a routine array that is interpreted by OUR code as a comment. Possibly the syntax would allow comments at the end of the strings you're proposing here, but if we don't want that level of parsing complexity just full like comments.

`"// Routine Comment"`

ArnoldSmith86 commented:
A `clickRoutine` entry like `"// comment"` already "works" and I'm not planning to change that.

A `"${a} = 1 + 2 // this is a comment"` was part of my proposal.

larkob commented:
I agree with the proposal, and just have one suggestion for improvement:

How about we avoid the use of curly braces on the left-hand side when they are used for variable assignment rather than for variable substitution? This is similar in Javascript (`var b = ´I am ${name}´ `).
```
"var b = 'p'",
"var a = 1 + 2 // this is a comment", // allow comments in the routine
"var d = property ${b} of ${idVar}", // the property name and widget ID come from variables
...
"var b.${var2} = 99.123", // the index taken from variable var2
```

The `var` keyword may be unnecessary, but I do think it helps to better convey the meaning of variable assignment.

ArnoldSmith86 commented:
I updated the title and description of this PR. It still needs a more complete guide of what's possible now. It's basically my proposal with @larkob's left-hand side. I also added a simple expression to test if a variable or property equals an integer (`${var == 1}`). Only equals and only positive integers (for now).

rogerl50 commented:
> 
>     * [ ]  If a `COMPUTE` uses a template in `applyVariables` that is no longer possible with the new syntax. The fileupdater should replace it by a series of `concat` or something like that. In practice no one uses this, I think.
> 
Why is template no longer possible? Can't I write
```
{
      "func": "SET",
      "property": "resultType",
      "collection": "thisButton",
      "value": "The result is ${selected}"
    },
```
I use the template functionality with `applyVariables` in several places.

ArnoldSmith86 commented:
That still works fine. But not for `COMPUTE` because it's gone.

larkob commented:
I've created a first draft of the documentation, which I have titled "Dynamic Expressions" - not sure if that's the most suitable name, let me know if you have better ideas.

https://github.com/ArnoldSmith86/virtualtabletop/wiki/Dynamic-Expressions-(WIP)

This documentation isn't targeted for beginners but for power users. I also wanted to document exceptions / edge cases as this will become helpful for automated testing. I also mention an "alternative syntax" where the operation keyword comes before the operations, which is currently not featured in PR #307, but is something I'm working on.

ArnoldSmith86 commented:
> I've created a first draft of the documentation

Very nice..

A `-` is also allowed in names. And you can use escapes like `\\u00a0`. Strings allow _a few_ more symbols. That should probably be mentioned.

"variable can be assigned to a value of any JSON type" - technically correct when array and object are added but it might suggest that this is closer to JSON than it is. Quotes are different, scientific notation for numbers is missing. Strings only allow very few symbols and escapes are different.

`var c = 'hello world' substr 1 5 // sets variable c to 'hello'` - shouldn't that be a 0?

Did you test all the stuff in Exception Handling? Without checking I would expect no 1 return values when there is no operation.

Missing things:
- `var $a = 1` which takes the variable name from variable `a`.
- `${x == 1}` which is mainly meant to be used in other operations.

ArnoldSmith86 commented:
In general this PR makes code _way_ more readable but if you really want to, you can construct challenges for the fileupdater:

Before:

```
"clickRoutine": [
  {
    "func": "COMPUTE",
    "operation": "concat",
    "operand1": "${asd}",
    "operand2": "{qwe}",
    "variable": "1 👍"
  },
  {
    "func": "COMPUTE",
    "operation": "concat",
    "applyVariables": [
      {
        "parameter": "operand1",
        "template": "${1 👍}{{1 👍}}"
      },
      {
        "parameter": "operand2",
        "template": "{}{}{}"
      }
    ],
    "variable": "2 👍"
  }
]
```

After:

```
"clickRoutine": [
  "var internal_computeMigration_existingVariable0 = '\\u0024\\u007basd\\u007d' // This was added by the automatic file migration because the new expression syntax does not support escaping variable expressions.",
  "var 1\\u0020\\ud83d\\udc4d = ${internal_computeMigration_existingVariable0} concat '\\u007bqwe\\u007d'",
  {
    "note": "This was added by the automatic file migration because the new expression syntax does not support templates in operands.",
    "func": "SET",
    "collection": "thisButton",
    "property": "internal_computeMigration_applyVariables1",
    "value": "{}{}{}"
  },
  {
    "note": "This was added by the automatic file migration because the new expression syntax does not support templates in operands.",
    "func": "SET",
    "collection": "thisButton",
    "property": "internal_computeMigration_applyVariables0",
    "value": "$${1\\u0020\\ud83d\\udc4d}${\\u007b1\\u0020\\ud83d\\udc4d}}"
  },
  "var 2\\u0020\\ud83d\\udc4d = ${PROPERTY internal_computeMigration_applyVariables0} concat ${PROPERTY internal_computeMigration_applyVariables1}",
  {
    "note": "This was added by the automatic file migration because the new expression syntax does not support templates in operands.",
    "func": "SET",
    "collection": "thisButton",
    "property": "internal_computeMigration_applyVariables1",
    "value": null
  },
  {
    "note": "This was added by the automatic file migration because the new expression syntax does not support templates in operands.",
    "func": "SET",
    "collection": "thisButton",
    "property": "internal_computeMigration_applyVariables0",
    "value": null
  }
]
```

96LawDawg commented:
You probably need to change the file updater "note" key to something other than "note."  Otherwise, won't it merge with the note already existing in that function?  I suggest "vtt_note" or "auto_note."

Also, I didn't think to ask this last time, but does the file updater include the files in the public library?  I know it will catch all the rooms on the server, but are the .vtt files in the public library included in that?

ArnoldSmith86 commented:
> You probably need to change the file updater "note" key to something other than "note." Otherwise, won't it merge with the note already existing in that function?

The `note`s you see in my previous comment are in operations that didn't exist in the original routine. They were inserted by the fileupdater from scratch so there's no way it merges with anything.

And the updater actually takes the parameter `note`, `Note`, `comment` or `Comment` from a `COMPUTE` it replaces and appends it as ` // <note>` to the resulting expression.

> does the file updater include the files in the public library?

Yes. Whenever a file loads it is sent through the updater. You can see that happening on the test server.

ArnoldSmith86 commented:
Unfortunately I had to make the migration of a `COMPUTE` with a `skip` a lot uglier. I hope that not too many people use that. I just couldn't find a way to make `IF` behave in the same way (or even another expression preceding the `IF`).

I'm not aware of any more issues now but I'm pretty sure there are still many left. The migration function in the fileupdater is now >160 lines long. It's always the details that kill you.

Please test this **a lot**! I would like to merge it next week if it's stable enough.

robartsd commented on client/js/widgets/widget.js:
```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
       }
 
+      if(typeof a == 'string') {
+        const identifier = '(?:[a-zA-Z0-9_-]|\\\\u[0-9a-fA-F]{4})+';
```
Identifier looks like a non-capturing group regular expression, but the others look like capturing groups. Also this expression allows identifiers to start with any character (or escape sequence): all numbers will match identifier.

ArnoldSmith86 replied:
True and true. I don't see a problem though. I'm pretty sure the only way this would be ambiguous is if we had an operation name that only contains digits. If you can provide an example that doesn't match correctly, I'm very interested.

robartsd commented on client/js/widgets/widget.js:
```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
+        const string     = `'((?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*)'`;
+        const number     = '(-?(?:0|[1-9][0-9]*)(?:\\.[0-9]+)?)';
+        const variable   = `(\\$\\{[^}]+\\})`;
+        const parameter  = `(null|true|false|\\[\\]|\\{\\}|${number}|${variable}|${string})`;
```
Empty arrays and objects can be parameters, but it looks like filled arrays and objects cannot.

ArnoldSmith86 replied:
True. That might be worth expanding in the future.

rogerl50 commented:
The IF statement
```
    {
      "func": "IF",
      "applyVariables": [
        {
          "parameter": "condition",
          "variable": "isCall"
        }
      ],
      "thenRoutine": [
        {
          "func": "LABEL",
          "collection": "lastBetPn",
          "applyVariables": [
            {
              "parameter": "value",
              "template": "Called {stringAmount}"
            }
          ],
          "Note": "Set last bet label from totalBet"
        }
      ],
....
```
is converted to
```
 {
      "func": "IF",
      "thenRoutine": [
        {
          "func": "LABEL",
          "collection": "lastBetPn",
          "applyVariables": [
            {
              "parameter": "value",
              "template": "Called {stringAmount}"
            }
          ],
          "Note": "Set last bet label from totalBet"
        }
      ],
... ,
      "condition": "${isCall}"
```
This fails because the inner `applyVariables` are not removed.

robartsd commented on client/js/widgets/widget.js:
```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
+        const variable   = `(\\$\\{[^}]+\\})`;
+        const parameter  = `(null|true|false|\\[\\]|\\{\\}|${number}|${variable}|${string})`;
+
+        const left       = `var (\\$)?(${identifier})(?:\\.(\\$)?(${identifier}))?`;
```
Allows for identifier.identifier, but not further levels (this is probably as intended).

robartsd commented on client/js/widgets/widget.js:
```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
+        const parameter  = `(null|true|false|\\[\\]|\\{\\}|${number}|${variable}|${string})`;
+
+        const left       = `var (\\$)?(${identifier})(?:\\.(\\$)?(${identifier}))?`;
+        const operation  = `${identifier}|[=+*/%<!>&|-]{1,2}`;
```
Will match some invalid pairs of characters from  `=+*/%<!>&|-`.

ArnoldSmith86 replied:
True but it'll report an invalid operation just like all the invalid `identifier`s.

96LawDawg commented:
So I was in the process of testing every public library game.  No issues on the first few games.  Got to Bug Chess and it breaks the room.  Nuked it.  Can't get anything else to work in the room.  Tested that 3 times in 3 rooms.  Definitely bug chess.  Which is strange because it doesn't have a `compute` any where in it.  Something else with the file updater?  I've suspended other public library testing for now.

robartsd commented on client/js/widgets/widget.js:
```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
+        const left       = `var (\\$)?(${identifier})(?:\\.(\\$)?(${identifier}))?`;
+        const operation  = `${identifier}|[=+*/%<!>&|-]{1,2}`;
+
+        const regex      = `^${left} = (?:${parameter}|(?:${parameter} )?(${operation})(?: ${parameter})?(?: ${parameter})?(?: ${parameter})?)( //.+)?`;
```
I don't particularly like the requirement that whitespace between "words" is exactly a single space character. I'd prefer matching `\s+` even though it looks uglier in code.

ArnoldSmith86 replied:
We can agree on `<space>+` but JSON doesn't allow tabs or newlines.

robartsd replied:
So other whitespace would need to be escaped in JSON. I guess ` +` makes the most sense then.

ArnoldSmith86 commented:
> Got to Bug Chess and it breaks the room.

Should be fixed. It contains this:

```JSON
{
  "func": "RECALL",
  "holder": "TheDeck",
  "applyVariables": [
    {
      "parameter": "owned",
      "variable": true
    }
  ]
}
```

ArnoldSmith86 commented:
> This fails because the inner `applyVariables` are not removed.

Should be fixed. Thank you **very** much for noticing. :D

robartsd commented on client/js/widgets/widget.js:
```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
+        const left       = `var (\\$)?(${identifier})(?:\\.(\\$)?(${identifier}))?`;
+        const operation  = `${identifier}|[=+*/%<!>&|-]{1,2}`;
+
+        const regex      = `^${left} += +(?:${parameter}|(?:${parameter} +)?(${operation})(?: +${parameter})?(?: +${parameter})?(?: +${parameter})?)( +//.+)?`;
```
Probably shouldn't capture whitespace before comment and empty comment will cause match to fail.
Replace last group of the expression with `(?: +(//.*))?`

ArnoldSmith86 replied:
The capture is not used anywhere.

Do we want empty comments though? Feels like a syntax error to me to write `var a = 1 //`. It's allowed in other languages I guess but what's the point?

robartsd replied:
I certainly wouldn't want to write code that has an empty comment, but I wouldn't want an empty comment to be a compile time error either.

robartsd commented on client/js/widgets/widget.js:
```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
+            }
+          };
+          const getValue = function(input) {
+            const toNum = s=>typeof s == 'string' && s.match(/^[-+]?[0-9]+(\.[0-9]+)?$/) ? +s : s;
```
Should this use the number pattern defined earlier?

ArnoldSmith86 replied:
In an ideal world, yes. This one has already been in the code for quite a while though and it turns numeric strings into numbers. I don't see a real reason to potentially break stuff by changing it (`"000"` would no longer be turned into `0`).

We could use this one in the expression parser but the other one should be more correct.

96LawDawg commented:
Tested all public library games successfully.

rogerl50 commented:
> > This fails because the inner `applyVariables` are not removed.
> 
> Should be fixed. Thank you **very** much for noticing. :D

It is fixed. My poker room appears to have been updated correctly. It has a fair amount of complicated code in it. If you want to examine the file update results, it is at http://212.47.248.129:3307/poker

ArnoldSmith86 commented:
> 1. After `a='abc'`, why does `$a=3` set the variable `abc` to `3`, but when `$a` is used on the right-hand side, `{}` are required, so `b = ${a}$`? What would `b=$a` do? How about `${a}=3`? What I'm asking is: do `$<var>` and `${<var>}` have meanings independent of the context in which they appear?

That's why my initial proposal was `${a} = 3` so it's always consistent but everyone disliked all those curly braces.

On the left it's currently always `var a =` or `var $a =` if the variable name comes from the variable `a`. On the right and in other operations it's always `${a}` or `${$a}` if the variable name comes from the variable `a`.

A `$a` on the right hand side is invalid syntax and should invalidate the whole expression.

So no, there's no different meaning. Just different ways to write it on the left and the right.

> 2. [...] Why can't it simply use [...]?

That's exactly what it did before you reported that it breaks for undefined variables. Trust me, I really don't like that it's _that_ ugly but I couldn't find a better way that behaves like `skip`.

But in the end it shouldn't really matter. Games that are still being worked on will just be prettified by the author (your solution is fine if the variable `positive` always contains `true` or `false`) and for other games no-one cares as long as it keeps working.

rogerl50 commented:
> On the left it's currently always `var a =` or `var $a =` if the variable name comes from the variable `a`. On the right and in other operations it's always `${a}` or `${$a}` if the variable name comes from the variable `a`.
> 
> A `$a` on the right hand side is invalid syntax and should invalidate the whole expression.

I know I'm late to the party here and should have paid attention before. But. I think this is error-prone. I know I will be writing $a on the right-hand side and cursing a lot :) I guess the {} are necessary on the right-hand side because there is potentially other syntax inside them.

Second question. You mention `${$a}`. After `var a='abc'`, `var abc='foo'`, does `b=${$a}` place the value of variable `foo` into variable `b`? Are these arbitrarily nestable?

ArnoldSmith86 commented:
> I think this is error-prone.

I'm open to changing this. Start a discussion with Raphael and Larkob. :smile:

> After `var a='abc'`, `var abc='foo'`, does `b=${$a}` place the value of variable `foo` into variable `b`?

It should. Not sure if anyone ever tested this.

> Are these arbitrarily nestable?

No. Just this one level.

larkob commented:
I just tested the variable substitution for index values and found a problem when the index is 0:

```
"var a = []",
"var a.0 = 2",
"var b = 0",
"var c = []",
"var c.0 = 4",
"var a.$b = ${c.$b}"
```
produces a debug log of:
```
  "a": 4,
  "b": 0,
  "c": [
    4
  ]
```

An index other than 0 is handled correctly:
``` 
"var a = []",
"var a.1 = 2",
"var b = 1",
"var c = []",
"var c.1 = 4",
"var a.$b = ${c.$b}"
```

produces a debug log of:
```
  "a": [
    null,
    4
  ],
  "b": 1,
  "c": [
    null,
    4
  ]
```

ArnoldSmith86 commented:
Can you add tests for that?

larkob commented:
I could, but I was wondering if we should even implement any index-related functionality on the left side at all. For now, all we aim for is a replacement of COMPUTE and applyVariables, Setting an index of a variable is not possible with applyVariables, only with the setIndex function, so all we needed is to support sth. like `set a = setIndex 1 'new entry'`. There's already a lot to review and consider, so I'd rather limit the scope to what is actually needed at the moment.

ArnoldSmith86 commented:
I don't agree. Because I think `getIndex` and `setIndex` are terrible and because it is already implemented.

Since the fileupdater doesn't use the indexing, it's not so bad if we miss a bug or two. So it's fine if it doesn't get reviewed as thoroughly. Obviously I want to fix the problem you found though.

ArnoldSmith86 commented:
> found a problem when the index is 0

Should be fixed now.

ArnoldSmith86 commented:
We noticed in today's meeting that it's no longer possible to read the operation from a variable (like the COMPUTE tutorial does).

You can now do this:

```Javascript
"clickRoutine": [
  "var operation = 'max'",
  "var test = 🧮operation 2 3"
]
```

ArnoldSmith86 commented:
Works for me: http://212.47.248.129:3307/8y34

It's an abacus: https://emojipedia.org/abacus/

I don't really think we should document the feature anywhere but the `COMPUTE` tutorial uses dynamic operations and I don't want to break that.

RaphaelAlvez commented:
@ArnoldSmith86 did you add the other trigonometric functions that were missing? should this go on another PR?

larkob commented:
I think this should go into another PR (I'm already preparing one)

robartsd commented:
> Not sure what this symbol "🧮" is supposed to be - anyhow it doesn't seem to work. But then again, I don't think we even need to support operations expressed as variables. Just because this was technically possible with applyVariables doesn't make it a requirement for the new approach in my opinion.

In the meeting I objected to crazy Unicode code points for this too (at first I thougt of `%`, but that is the remainder operator - `#` is a symbol in the ASCII code page that could be used though). This particular symbol does not display properly on some of my systems; but I don't care that much since I never plan to use this as a feature. (I care more about the use of these types of symbols in other places - sometimes the JSON editor menu has to be operated based on my memory of the positions of the buttons due to missing symbols on my systems.)

ArnoldSmith86 commented:
OK, back to "everything I'm aware of is done".

The latest commit takes care of migrating stuff like this:

```Javascript
{
  "func": "COMPUTE",
  "applyVariables": [
    {
      "parameter": "variable",
      "template": "{templateOne}?{templateTwo}"
    },
    {
      "parameter": "operation",
      "template": "{opTempOne}in"
    }
  ],
  "operand1": 7,
  "operand2": 3
},
{
  "func": "COMPUTE",
  "applyVariables": [
    {
      "parameter": "variable",
      "property": "varThree",
      "widget": " OF "
    },
    {
      "parameter": "operation",
      "property": "opThree"
    }
  ],
  "operand1": 2,
  "operand2": 3
}
```

You can look at all the weird stuff I tested here: https://virtualtabletop.io/f4yv
The same file is added here: http://212.47.248.129:3307/pr-test

rogerl50 commented:
A new upload/file update of my poker room seems to be fully functional ( http://212.47.248.129:3307/poker if you want to play with it.)

rogerl50 reviewed (approved):
A new upload/file update of my poker room seems to be fully functional ( http://212.47.248.129:3307/poker if you want to play with it). Similarly with Dots, which also has some pretty complex code in it.

Approved through test.

96LawDawg reviewed (approved):
I've tested it some more.  I don't see any issues.  Recommend merge.

RaphaelAlvez reviewed (approved):
Approved by review of other people's Routines.

ArnoldSmith86 commented:
I just updated https://virtualtabletop.io/CitiesGenerator after merging.

I manually updated the routines to take advantage of the new possibilities.

Before:

```JSON
{
  "previewRoutine": [
    {
      "func": "COMPUTE",
      "operation": "getIndex",
      "operand1": [
        0,
        {}
      ]
    },
    {
      "func": "COMPUTE",
      "operation": "getIndex",
      "operand1": [
        0,
        {}
      ],
      "variable": "cardTypes"
    },
    {
      "func": "SELECT",
      "property": "id",
      "value": "result"
    },
    {
      "func": "GET",
      "property": "text",
      "variable": "result"
    },
    {
      "func": "GET",
      "property": "text",
      "collection": "thisButton"
    },
    {
      "func": "SELECT",
      "property": "id",
      "value": "rotationButton"
    },
    {
      "func": "GET",
      "property": "rIndex"
    },
    {
      "func": "GET",
      "property": "xIndex",
      "collection": "thisButton"
    },
    {
      "func": "GET",
      "property": "yIndex",
      "collection": "thisButton"
    },
    {
      "func": "COMPUTE",
      "operation": "setIndex",
      "operand1": "css",
      "applyVariables": [
        {
          "parameter": "operand2",
          "template": "{result};--x{text}:{xIndex};--y{text}:{yIndex};--r{text}:{rIndex}"
        }
      ]
    },
    {
      "func": "COMPUTE",
      "operation": "setIndex",
      "operand1": "preview",
      "applyVariables": [
        {
          "parameter": "operand2",
          "variable": "COMPUTE"
        }
      ],
      "variable": "cardTypes"
    },
    {
      "func": "SELECT",
      "property": "id",
      "value": "bo9g"
    },
    {
      "func": "SET",
      "property": "cardTypes",
      "applyVariables": [
        {
          "parameter": "value",
          "variable": "cardTypes"
        }
      ]
    }
  ]
```

After:

```JSON
{
  "previewRoutine": [
    {
      "func": "SELECT",
      "property": "id",
      "value": "bo9g"
    },
    {
      "func": "SET",
      "property": "cardTypes",
      "value": {
        "preview": {
          "css": "${PROPERTY text OF result};--x${PROPERTY text}:${PROPERTY xIndex};--y${PROPERTY text}:${PROPERTY yIndex};--r${PROPERTY text}:${PROPERTY rIndex OF rotationButton}"
        }
      }
    }
  ]
}
```
